man/ali.man
.TH ALI %manext1% 2013-12-05 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
ali \- list nmh mail aliases
.SH SYNOPSIS
.HP 5
.na
.B ali
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-alias
.IR aliasfile ]
.RB [ \-list " | " \-nolist ]
.RB [ \-user " | " \-nouser ]
.RI [ aliases " ...]"
.ad
.SH DESCRIPTION
.B ali
searches the named mail alias files for each of the given
.IR aliases .
It creates a list of addresses for those
.IR aliases ,
and writes that list on standard output.  If no arguments are given,
.B ali
outputs all alias entries.
.PP
By default, when an alias expands to multiple addresses, the addresses
are separated by commas and printed on as few lines as possible.  If the
.B \-list
switch is specified, and an address expands to multiple
addresses, each address will appear on a separate line.
.PP
The
.B \-user
switch directs
.B ali
to perform its processing in
an inverted fashion: instead of listing the addresses that each given
alias expands to,
.B ali
will list the aliases that expand to each
given address.
.PP
The files specified by the profile entry
.RI \*(lq Aliasfile \*(rq
and any additional alias files given by the
.B \-alias
.I aliasfile
switch will be read.  Each
.I alias
is processed as described in
.IR mh\-alias (5).
.SH FILES
Alias files are looked up in multiple locations: if the pathname is
absolute or starts with
.RI \*(lq . \*(rq
or
.RI \*(lq .. \*(rq
it is accessed directly; otherwise tilde expansion is done on
usernames, then files are searched for in the user's
.I Mail
directory as specified in their profile.  If not found there, the directory
.RI \*(lq %nmhetcdir% \*(rq
is checked.
.PP
.TP 20
$HOME/.mh_profile
The user's profile.
.SH "PROFILE COMPONENTS"
.PP
.PD 0
.TP 20
Path:
To determine the user's nmh directory
.TP
Aliasfile:
For a default alias file
.PD
.SH "SEE ALSO"
.IR mh\-alias (5)
.SH DEFAULTS
.PD 0
.TP 20
aliasfile
%nmhetcdir%/MailAliases
.TP
\-nolist
.TP
\-nouser
.PD
.SH CONTEXT
None


NEWS
#
# NEWS - Release notes for nmh 1.6
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly two years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.6 release of nmh:

- All nmh programs will now warn about multiple profile entries for
  the same non-null, non-comment component.  In previous versions, all
  but the first were silently ignored.
- mhmail now supports all post(8) options, or optionally can use
  send(1) with all of its options.  Its formerly undocumented -resent
  switch has been documented.  And it has new -attach, -headerfield,
  and -send/-nosend switches.
- Default to flock() locking on OpenBSD and Darwin.
- Added -directives support to mhbuild(1) to control whether or not
  mhbuild will honor MIME directives by default.  And added support
  for special #on/#off/#pop directives to control the MIME directive
  processing state.
- Added -messageid switch to send(1) and post(8).  This allows
  selection of the style to use for generated Message-ID and
  Content-ID header fields.  The default localname style is
  pid.time@localname, where time is in seconds, and matches previous
  behavior.  The random style replaces the localname with some
  (pseudo)random bytes and uses microsecond-resolution time.
- Added -clobber switch to mhstore(1) to control overwriting of
  existing files.
- Added -outfile switch to mhstore(1).
- Added -noall/-all switches to sortm(1).  sortm -noall requires
  a messages argument.
- $PAGER overrides the compiled-in default pager command.
- Added etc/scan.highlighted format file, as an example of how to
  highlight/colorize the output of scan(1).
- inc(1) now supports a -port switch to specify the port used by the
  POP server.
- pick(1) now decodes MIME-encoded header fields before searching.
- The VISUAL and EDITOR environment variables are now supported as fallbacks
  if the user does not configure an editor entry in their profile.
- The format engine (mh_format(5)) now properly accounts for multibyte
  characters when accounting for column widths.
- burst(1) now can burst MIME-formatted digests (messages that contain
  message/rfc822 parts instead of messages formatted with RFC 934).
- All proc entries (showproc, moreproc, etc) can now accept entries that
  contain spaces and shell metacharacters.  If found, such entries will
  either be space-split or processed by /bin/sh.
- A new program, fmttest(1) is included to help debug format files.  See
  NOTABLE CHANGES below.
- mhshow/mhstore now have support for RFC-2017 (access-type=url) for
  external message bodies.
- Added -retainsequences switch to refile(1).
- A new program, mhfixmsg(1), is included to rewrite MIME messages with
  various transformations.  See NOTABLE CHANGES below.
- Added -[no]rmmproc switches to rmm(1).
- Added support for Content-Disposition header (RFC 2183) to mhstore(1)
  and mhn(1) when used with -auto.
- All nmh commands now support transactional locking for sequence files.
  See NOTABLE CHANGES below.
- There is no longer a per-folder maximum number of sequences.
- For the SMTP MTA TLS can now be negotiated at the beginning of the
  connection with the -initialtls switch.
- Messages can now be selected using a relative offset from some other
  message, or from the start or end of a sequence.  See mh-sequence(5).
- The -changecur and -nochangecur switches have been added to mhlist(1).
- mhbuild(1) can now encode 8-bit message headers using RFC-2047 encoding
  rules.  See NOTABLE CHANGES below.
- The whatnow(1) attach feature will determine the content type of an
  attachment using a program such as file --mime-type, if available at
  configuration time.  If not, it will use mhshow-suffix- entries as
  before.  The -v switch to attach causes it to display the mhbuild
  directive that send(1) will use.
- mhbuild(1) now supports the -auto/-noauto flags (to be used by send(1)
  when invoking mhbuild automatically).
- mhbuild(1) now is automatically run by send, to insure that all outgoing
  messages have proper MIME formatting.  See NOTABLE CHANGES below.
- A new header, "Attach", is supported by mhbuild; it is used to replace
  previous functionality (which by default used a header named
  Nmh-Attachment).  See NOTABLE CHANGES below.
- The default Content-Transfer-Encoding for text parts is now 8bit.
- mhbuild(1) now supports a selectable Content-Transfer-Encoding
- If nmh was configured with iconv(3) support, mhshow will convert, if
  necessary, the charset of text/plain content to match the user's
  locale setting.  See NOTABLE CHANGES below.
- Added support for %{charset} display string escape to mhshow(1).
- The MIME parsing and generating routines now support RFC 2231 extended
  parameter information.
- mh-mime(7) now provides an introduction to nmh's MIME handling.
- mhshow(1) will now by default display all text content under one pager,
  and display markers for non-text and non-inline content.  The content
  markers are changeable via mh-format(5).  See NOTABLE CHANGES below.

-----------------
OBSOLETE FEATURES
-----------------
- Changed exit status of each nmh command's -version and -help
  switches from 1 to 0.
- The following environment variables were deprecated in nmh 1.5
  and removed from this release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature removed)
- Changed repl and dist default to -noatfile.  The default of -atfile

uip/ali.c
/*
 * ali.c -- list nmh mail aliases
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/addrsbr.h>
#include <h/aliasbr.h>
#include <h/mts.h>
#include <h/utils.h>

#define ALI_SWITCHES \
    X("alias aliasfile", 0, ALIASW) \
    X("noalias", 0, NALIASW) \
    X("list", 0, LISTSW) \
    X("nolist", 0, NLISTSW) \
    X("user", 0, USERSW) \
    X("nouser", 0, NUSERSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(ALI);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(ALI, switches);
#undef X

static int pos = 1;

extern struct aka *akahead;

/*
 * prototypes
 */
static void print_aka (char *, int, int);
static void print_usr (char *, int);


int
main (int argc, char **argv)
{
    int i, vecp = 0, inverted = 0, list = 0;
    int noalias = 0;
    char *cp, **ap, **argp, buf[BUFSIZ];
    /* Really only need to allocate for argc-1, but must allocate at least 1,
       so go ahead and allocate for argc char pointers. */
    char **vec = mh_xmalloc (argc * sizeof (char *)), **arguments;
    struct aka *ak;

    if (nmh_init(argv[0], 1)) { return 1; }

    mts_init ();
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [switches] aliases ...",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version (invo_name);
		    done (0);

		case ALIASW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if ((i = alias (cp)) != AK_OK)
			adios (NULL, "aliasing error in %s - %s", cp, akerror (i));
		    continue;
		case NALIASW: 
		    noalias++;
		    continue;

		case LISTSW: 
		    list++;
		    continue;
		case NLISTSW: 
		    list = 0;
		    continue;

		case USERSW: 
		    inverted++;
		    continue;
		case NUSERSW: 
		    inverted = 0;
		    continue;
	    }
	}

	if (vecp < argc) {
	    vec[vecp++] = cp;
	} else {
	    /* Should never happen, but try to protect against code changes
	       that could allow it. */
	    adios (NULL, "too many arguments");
	}
    }

    if (!noalias) {
	/* allow Aliasfile: profile entry */
	if ((cp = context_find ("Aliasfile"))) {
	    char *dp = NULL;

	    for (ap = brkstring(dp = mh_xstrdup(cp), " ", "\n"); ap && *ap; ap++)
		if ((i = alias (*ap)) != AK_OK)
		    adios (NULL, "aliasing error in %s - %s", *ap, akerror (i));
            mh_xfree(dp);
	}
	alias (AliasFile);
    }

    /*
     * If -user is specified
     */
    if (inverted) {
	if (vecp == 0)
	    adios (NULL, "usage: %s -user addresses ...  (you forgot the addresses)",
		   invo_name);

	for (i = 0; i < vecp; i++)
	    print_usr (vec[i], list);
    } else {
	if (vecp) {
	    /* print specified aliases */
	    for (i = 0; i < vecp; i++)
		print_aka (akvalue (vec[i]), list, 0);
	} else {
	    /* print them all */
	    for (ak = akahead; ak; ak = ak->ak_next) {
                char *res;

		printf ("%s: ", ak->ak_name);
		pos += strlen (ak->ak_name) + 1;
                res = akresult(ak);
		print_aka(res, list, pos);
                free(res);
	    }
	}
    }

    free (vec);
    done (0);
    return 1;
}

static void
print_aka (char *p, int list, int margin)
{
    char c;

    if (p == NULL) {
	puts("<empty>");
	return;
    }

    while ((c = *p++)) {
	switch (c) {
	    case ',': 
		if (*p) {
		    if (list)
			printf ("\n%*s", margin, "");
		    else {
			if (pos >= 68) {
			    printf (",\n ");
			    pos = 2;
			} else {
			    printf (", ");
			    pos += 2;
			}
		    }
		}

	    case 0: 
		break;

	    default: 
		pos++;
		putchar (c);
	}
    }

    putchar ('\n');
    pos = 1;
}

static void
print_usr (char *s, int list)
{
    char *cp, *pp, *vp;
    struct aka *ak;
    struct mailname *mp, *np;

    if ((pp = getname (s)) == NULL)
	adios (NULL, "no address in \"%s\"", s);
    if ((mp = getm (pp, NULL, 0, NULL, 0)) == NULL)
	adios (NULL, "bad address \"%s\"", s);
    while (getname (""))
	continue;

    vp = NULL;
    for (ak = akahead; ak; ak = ak->ak_next) {
	pp = akresult (ak);
	while ((cp = getname (pp))) {
	    if ((np = getm (cp, NULL, 0, NULL, 0)) == NULL)
		continue;
	    if (!strcasecmp (mp->m_host ? mp->m_host : "",
			     np->m_host ? np->m_host : "")  &&
		!strcasecmp (mp->m_mbox ? mp->m_mbox : "",
			     np->m_mbox ? np->m_mbox : "")) {
		vp = vp ? add (ak->ak_name, add (",", vp))
		    : getcpy (ak->ak_name);
		mnfree (np);
		while (getname (""))
		    continue;
		break;
	    }
	    mnfree (np);
	}
    }
    mnfree (mp);

    print_aka (vp ? vp : s, list, 0);

    mh_xfree(vp);
}
